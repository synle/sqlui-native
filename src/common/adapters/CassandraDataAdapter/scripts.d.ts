import BaseDataScript from 'src/common/adapters/BaseDataAdapter/scripts';
import { SqlAction } from 'typings';
export declare function getSelectAllColumns(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getSelectSpecificColumns(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getInsert(input: SqlAction.TableInput, value?: Record<string, any>): SqlAction.Output | undefined;
export declare function getBulkInsert(input: SqlAction.TableInput, rows?: Record<string, any>[]): SqlAction.Output | undefined;
export declare function getUpdateWithValues(input: SqlAction.TableInput, value: Record<string, any>, conditions: Record<string, any>): SqlAction.Output | undefined;
export declare function getUpdate(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getDelete(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getCreateKeyspace(input: SqlAction.DatabaseInput): SqlAction.Output | undefined;
export declare function getDropKeyspace(input: SqlAction.DatabaseInput): SqlAction.Output | undefined;
export declare function getCreateConnectionDatabase(input: SqlAction.ConnectionInput): SqlAction.Output | undefined;
export declare function getCreateTable(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getDropTable(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getAddColumn(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getDropColumns(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare class ConcreteDataScripts extends BaseDataScript {
    dialects: string[];
    getIsTableIdRequiredForQuery(): boolean;
    getSyntaxMode(): string;
    supportMigration(): boolean;
    supportCreateRecordForm(): boolean;
    supportEditRecordForm(): boolean;
    getTableScripts(): (typeof getInsert)[];
    getDatabaseScripts(): (typeof getCreateKeyspace)[];
    getConnectionScripts(): (typeof getCreateConnectionDatabase)[];
    getSampleConnectionString(dialect: any): string;
    getSampleSelectQuery(tableActionInput: any): SqlAction.Output;
    getCodeSnippet(connection: any, query: any, language: any): string;
}
declare const _default: ConcreteDataScripts;
export default _default;
