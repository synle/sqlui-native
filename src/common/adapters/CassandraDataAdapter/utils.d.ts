declare type ClientOptions = {
    host: string;
    port: number;
    contactPoints: string[];
    keyspace?: string;
    authProvider?: {
        username: string;
        password: string;
    };
};
export declare function getClientOptions(connectionOption: string, database?: string): ClientOptions;
export {};
