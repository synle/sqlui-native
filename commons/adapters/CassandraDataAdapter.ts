import * as cassandra from 'cassandra-driver';
import { SqluiCore } from '../../typings';
import IDataAdapter from './IDataAdapter';
import BaseDataAdapter from './BaseDataAdapter';

export default class CassandraAdapter extends BaseDataAdapter implements IDataAdapter {
  dialect: SqluiCore.Dialect = 'cassandra';
  /**
   * cassandra version
   * @type {number}
   */
  version?: string;
  isCassandra2?: boolean;

  constructor(connectionOption: string) {
    super(connectionOption);
    this.dialect = 'cassandra';
  }

  private async getConnection(database?: string): Promise<cassandra.Client> {
    // attempt to pull in connections
    return new Promise<cassandra.Client>(async (resolve, reject) => {
      try {
        const connectionParameters = BaseDataAdapter.getConnectionParameters(this.connectionOption);

        const connectionHosts = connectionParameters?.hosts || [];
        if(connectionHosts.length === 0){
          // we need a host in the connection string
          reject('Invalid connection. Host and Port not found');
        }

        //@ts-ignore
        const client = new cassandra.Client({
          contactPoints: [connectionHosts[0].host],
          protocolOptions: {
            port: connectionHosts[0].port,
          },
          keyspace: database,
        });

        await this.authenticate(client);

        resolve(client);
      } catch (err) {
        reject(err);
      }
    });
  }

  async authenticate(client?: cassandra.Client) {
    // TODO: To Be Implemented
    return new Promise<void>((resolve, reject) => {
      client?.connect((err: unknown) => {
        if (err) {
          client?.shutdown();
          return reject(err);
        }

        this.version = client?.getState().getConnectedHosts()[0].cassandraVersion;
        this.isCassandra2 = this.version === '2';

        resolve();
      });
    });
  }

  async getDatabases(): Promise<SqluiCore.DatabaseMetaData[]> {
    const client = await this.getConnection();

    //@ts-ignore
    const keyspaces = Object.keys(client?.metadata?.keyspaces);
    return keyspaces
      .map((keyspace) => ({
        name: keyspace,
        tables: [],
      }))
      .sort((a, b) => (a.name || '').localeCompare(b.name || ''));
  }

  async getTables(database?: string): Promise<SqluiCore.TableMetaData[]> {
    if (!database) {
      return [];
    }

    let sql;
    if (this.isCassandra2 === true) {
      sql = `
          SELECT columnfamily_name as name
          FROM system.schema_columnfamilies
          WHERE keyspace_name = ?
        `;
    } else {
      sql = `
          SELECT table_name as name
          FROM system_schema.tables
          WHERE keyspace_name = ?
        `;
    }

    const res = await this._execute(sql, [database]);

    return res.rows
      .map((row) => ({
        name: row.name,
        columns: [],
      }))
      .sort((a, b) => (a.name || '').localeCompare(b.name || ''));
  }

  async getColumns(table: string, database?: string): Promise<SqluiCore.ColumnMetaData[]> {
    if (!database) {
      return [];
    }

    let sql;
    if (this.isCassandra2 === true) {
      sql = `
        SELECT type as position, column_name, validator as type
        FROM system.schema_columns
        WHERE keyspace_name = ?
          AND columnfamily_name = ?
      `;
    } else {
      sql = `
        SELECT position, column_name, type
        FROM system_schema.columns
        WHERE keyspace_name = ?
          AND table_name = ?
      `;
    }
    const res = await this._execute(sql, [database, table]);

    return res.rows
      .map((row) => ({
        name: row.column_name,
        type: row.type,
      }))
      .sort((a, b) => (a.name || '').localeCompare(b.name || ''));
  }

  private async _execute(sql: string, params?: string[], database?: string) {
    const client = await this.getConnection(database);
    try {
      let res = await client.execute(sql, params || [], { prepare: true });
      client?.shutdown();
      return res;
    } catch (err) {
      client?.shutdown();
      throw err;
    }
  }

  async execute(sql: string, database?: string): Promise<SqluiCore.Result> {
    let rawToUse: any | undefined;
    let metaToUse: any | undefined;

    try {
      const res = await this._execute(sql, undefined, database);

      if (!res) {
        return {
          ok: false,
          raw: [],
          meta: null,
        };
      }
      const { rows } = res;

      //@ts-ignore
      delete res.rows;

      rawToUse = rows;
      metaToUse = res;

      return {
        ok: true,
        raw: rawToUse,
        meta: metaToUse,
      };
    } catch (error) {
      return {
        ok: false,
        error,
      };
    }
  }
}
