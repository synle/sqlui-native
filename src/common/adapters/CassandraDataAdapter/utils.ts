import BaseDataAdapter from 'src/common/adapters/BaseDataAdapter/index';

type ClientOptions = {
  host: string;
  port: number;
  contactPoints: string[];
  keyspace?: string;
  authProvider?: {
    username: string;
    password: string;
  };
};

export function getClientOptions(connectionOption: string, database?: string): ClientOptions {
  const connectionParameters = BaseDataAdapter.getConnectionParameters(connectionOption);

  const connectionHosts = connectionParameters?.hosts || [];
  if (connectionHosts.length === 0) {
    // we need a host in the connection string
    throw 'Invalid connection. Host and Port not found';
  }

  const host = connectionHosts[0].host;
  const port = connectionHosts[0].port || 9042;

  const clientOptions: ClientOptions = {
    host,
    port,
    contactPoints: [`${host}:${port}`],
  };

  if (database) {
    clientOptions.keyspace = database;
  }

  // client authentication
  if (connectionParameters?.username && connectionParameters?.password) {
    clientOptions.authProvider = {
      username: connectionParameters?.username,
      password: connectionParameters?.password,
    };
  }

  return clientOptions;
}