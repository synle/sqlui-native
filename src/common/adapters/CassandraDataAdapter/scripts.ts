import BaseDataScript, { getDivider } from 'src/common/adapters/BaseDataAdapter/scripts';
import { getClientOptions } from 'src/common/adapters/CassandraDataAdapter/utils';
import { escapeSQLValue, isValueBoolean, isValueNumber } from 'src/frontend/utils/formatter';
import { SqlAction, SqluiCore } from 'typings';

const formatter = 'sql';

function _isColumnNumberField(col: SqluiCore.ColumnMetaData) {
  return col.type.toLowerCase().includes('int') || col.type.toLowerCase().includes('float');
}

function _isColumnBooleanField(col: SqluiCore.ColumnMetaData) {
  return col.type.toLowerCase().includes('boolean');
}

function _getDummyColumnValue(col: SqluiCore.ColumnMetaData) {
  if (_isColumnBooleanField(col)) {
    return 'true';
  } else if (_isColumnNumberField(col)) {
    return '123';
  } else {
    // other types need to be wrapped in single quote
    return `'_${col.name}_'`;
  }
}

export function getSelectAllColumns(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Select All Columns`;

  return {
    label,
    formatter,
    query: `SELECT *
              FROM ${input.tableId}
              LIMIT ${input.querySize}`,
  };
}

export function getSelectSpecificColumns(
  input: SqlAction.TableInput,
): SqlAction.Output | undefined {
  const label = `Select Specific Columns`;

  if (!input.columns) {
    return undefined;
  }

  const columnString = `\n` + input.columns.map((col) => `  ${col.name}`).join(',\n');
  const whereColumnString = input.columns.map((col) => `${col.name} = ''`).join('\n AND ');

  return {
    label,
    formatter,
    query: `SELECT ${columnString}
              FROM ${input.tableId}
              WHERE ${whereColumnString}
              LIMIT ${input.querySize}`,
  };
}

export function getInsert(
  input: SqlAction.TableInput,
  value?: Record<string, any>,
): SqlAction.Output | undefined {
  const label = `Insert`;

  if (!input.columns) {
    return undefined;
  }

  const columnString = input.columns.map((col) => col.name).join(',\n');
  const insertValueString = input.columns.map((col) => {
    let valToUse: string | null = '';

    if (value) {
      if (value?.[col.name] === null) {
        valToUse = null;
      } else if (value?.[col.name] !== undefined) {
        // use the value if it's there
        valToUse = `${escapeSQLValue(value[col.name])}`;
      }

      if (valToUse === undefined) {
        valToUse = null;
      }

      if (valToUse === null || valToUse === 'null') {
        return 'null';
      }

      if (_isColumnBooleanField(col)) {
        valToUse = (valToUse || '').toLowerCase();
        if (valToUse === 'true' || valToUse === '1') {
          return 'true';
        }

        if (valToUse === 'false' || valToUse === '0') {
          return 'false';
        }

        return 'null'; // no value, then returned as null
      }

      if (_isColumnNumberField(col)) {
        // don't wrap with quote
        return valToUse;
      }

      return `'${valToUse}'`;
    } else {
      // no value, generate dummy data
      return _getDummyColumnValue(col);
    }
  });

  return {
    label,
    formatter,
    query: `INSERT INTO ${input.tableId} (${columnString})
              VALUES (${insertValueString})`,
  };
}

export function getBulkInsert(
  input: SqlAction.TableInput,
  rows?: Record<string, any>[],
): SqlAction.Output | undefined {
  const label = `Insert`;

  if (!input.columns) {
    return undefined;
  }

  if (!rows || rows.length === 0) {
    return undefined;
  }

  const rowsToInsert = (rows || [])
    .map((value) => getInsert(input, value))
    .map((output) => output?.query);

  return {
    label,
    formatter,
    query: `
    BEGIN BATCH
    ${rowsToInsert.join(';\n')}
    APPLY BATCH
    `,
  };
}

export function getUpdateWithValues(
  input: SqlAction.TableInput,
  value: Record<string, any>,
  conditions: Record<string, any>,
): SqlAction.Output | undefined {
  const label = `Update`;

  if (!input.columns) {
    return undefined;
  }

  const columnString = Object.keys(value)
    .map((colName) => {
      let valToUse = value[colName];

      if (!isValueNumber(valToUse) && !isValueBoolean(valToUse)) {
        // wrap the single quote for string
        valToUse = `'${escapeSQLValue(valToUse)}'`;
      }

      return `${colName} = ${valToUse}`;
    })
    .join(', \n');

  const whereColumnString = Object.keys(conditions)
    .map((colName) => {
      let valToUse = conditions[colName];

      if (!isValueNumber(valToUse)) {
        // wrap the single quote for string
        valToUse = `'${escapeSQLValue(valToUse)}'`;
      }

      return `${colName} = ${valToUse}`;
    })
    .join(' AND \n');

  return {
    label,
    formatter,
    query: `UPDATE ${input.tableId}
                SET ${columnString}
                WHERE ${whereColumnString}`,
  };
}

export function getUpdate(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Update`;

  if (!input.columns) {
    return undefined;
  }

  const columnString = input.columns
    .map((col) => `${col.name} = ${_getDummyColumnValue(col)}`)
    .join(',\n');
  const whereColumnString = input.columns
    .map((col) => `${col.name} = ${_getDummyColumnValue(col)}`)
    .join(' AND \n');

  return {
    label,
    formatter,
    query: `UPDATE ${input.tableId}
              SET ${columnString}
              WHERE ${whereColumnString}`,
  };
}

export function getDelete(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Delete`;

  if (!input.columns) {
    return undefined;
  }

  const whereColumnString = input.columns.map((col) => `${col.name} = ''`).join(' AND \n');

  return {
    label,
    formatter,
    query: `DELETE FROM ${input.tableId}
              WHERE ${whereColumnString}`,
  };
}

export function getCreateKeyspace(input: SqlAction.DatabaseInput): SqlAction.Output | undefined {
  const label = `Create Keyspace`;

  return {
    label,
    formatter,
    query: `CREATE KEYSPACE IF NOT EXISTS ${input.databaseId || 'some_keyspace'}
             WITH replication = {'class': 'SimpleStrategy', 'replication_factor' : 3};`,
  };
}

export function getDropKeyspace(input: SqlAction.DatabaseInput): SqlAction.Output | undefined {
  const label = `Drop Keyspace`;

  return {
    label,
    formatter,
    query: `DROP KEYSPACE IF EXISTS ${input.databaseId};`,
  };
}

export function getCreateConnectionDatabase(
  input: SqlAction.ConnectionInput,
): SqlAction.Output | undefined {
  const label = `Create Connection Keyspace`;

  return {
    label,
    formatter,
    query: `CREATE KEYSPACE IF NOT EXISTS some_keyspace
           WITH replication = {'class': 'SimpleStrategy', 'replication_factor' : 3};`,
  };
}

export function getCreateTable(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Create Table`;

  if (!input.columns) {
    return undefined;
  }

  let columnString: string = '';
  // mapping column
  columnString = input.columns
    .map((col) => [col.name, col.type, col.primaryKey ? 'PRIMARY KEY' : ''].join(' '))
    .join(',\n');

  // figuring out the keys
  let partitionKeys: string[] = [],
    clusteringKeys: string[] = [];
  for (const col of input.columns) {
    if (col.kind === 'partition_key') {
      partitionKeys.push(col.name);
    } else if (col.kind === 'clustering') {
      clusteringKeys.push(col.name);
    }
  }
  if (partitionKeys.length > 0) {
    if (clusteringKeys.length > 0) {
      columnString += `, PRIMARY KEY((${partitionKeys.join(', ')}), ${clusteringKeys.join(', ')})`;
    } else {
      // has only the partition key
      columnString += `, PRIMARY KEY((${partitionKeys.join(', ')}))`;
    }
  }

  return {
    label,
    formatter,
    query: `CREATE TABLE ${input.tableId} (${columnString})`,
  };
}

export function getDropTable(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Drop Table`;

  return {
    label,
    formatter,
    query: `DROP TABLE ${input.tableId}`,
  };
}

export function getAddColumn(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Add Column`;

  return {
    label,
    formatter,
    query: `ALTER TABLE ${input.tableId}
            ADD new_column1 TEXT`,
  };
}

export function getDropColumns(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Drop Column`;

  return {
    label,
    formatter,
    query: input.columns
      ?.map(
        (col) => `ALTER TABLE ${input.tableId}
                     DROP ${col.name};`,
      )
      .join('\n'),
  };
}

export class ConcreteDataScripts extends BaseDataScript {
  dialects = ['cassandra'];

  getIsTableIdRequiredForQuery() {
    return false;
  }

  getSyntaxMode() {
    return 'sql';
  }

  supportMigration() {
    return true;
  }

  supportCreateRecordForm() {
    return true;
  }

  supportEditRecordForm() {
    return true;
  }

  getTableScripts() {
    return [
      getSelectAllColumns,
      getSelectSpecificColumns,
      getInsert,
      getDivider,
      getUpdate,
      getDelete,
      getDivider,
      getCreateTable,
      getDropTable,
      getAddColumn,
      getDropColumns,
    ];
  }

  getDatabaseScripts() {
    return [
      getDivider,
      getCreateKeyspace, // TODO: right now this command does not tie to the input, it will hard code the keyspace to be some_keyspace
      getDropKeyspace,
    ];
  }

  getConnectionScripts() {
    return [getDivider, getCreateConnectionDatabase];
  }

  getSampleConnectionString(dialect) {
    return `cassandra://username:password@localhost:9042`;
  }

  getSampleSelectQuery(actionInput: SqlAction.TableInput) {
    return getSelectAllColumns(actionInput);
  }

  getCodeSnippet(
    connection: SqluiCore.ConnectionProps,
    query: SqluiCore.ConnectionQuery,
    language: SqluiCore.LanguageMode,
  ) {
    const sql = query.sql;
    const database = query.databaseId;
    const clientOptions = getClientOptions(connection.connection, database);

    switch (language) {
      case 'javascript':
        return `
// npm install --save cassandra-driver
const cassandra = require('cassandra-driver')

async function _doWork(){
  try {
    const clientOptions = ${JSON.stringify(clientOptions)};
    const client = new cassandra.Client({
      contactPoints: clientOptions.contactPoints,
      keyspace: clientOptions.keyspace,
      authProvider: clientOptions.authProvider ? new cassandra.auth.PlainTextAuthProvider(
        clientOptions.authProvider.username,
        clientOptions.authProvider.password,
      ): undefined,
      sslOptions: {
        rejectUnauthorized: false, // optional, check to see if you need to disable this SSL check
      }
    });
    await new Promise((resolve, reject) => {
      client.connect((err) => err ? reject(err): resolve());
    })

    const res = await client.execute(\`${sql}\`);
    console.log(res);
  } catch(err){
    console.log('Failed to connect', err);
  }
}

_doWork();
        `.trim();

      case 'python':
        return `
# python3 -m venv ./ # setting up virtual environment with
# source bin/activate # activate the venv profile
# pip install cassandra-driver
from cassandra.cluster import Cluster
from ssl import PROTOCOL_TLSv1_2, SSLContext, CERT_NONE
from cassandra.auth import PlainTextAuthProvider

# remove \`ssl_context=ssl_context\` to disable SSL (applicable for Cassandra in CosmosDB)
ssl_context = SSLContext(PROTOCOL_TLSv1_2)
ssl_context.verify_mode = CERT_NONE
cluster = Cluster(['${clientOptions.host}'], port=${
          clientOptions.port
        }, auth_provider=PlainTextAuthProvider(username='${
          clientOptions?.authProvider?.username || ''
        }', password='${clientOptions?.authProvider?.password || ''}'), ssl_context=ssl_context)
session = cluster.connect()

session.execute('USE ${database || 'some_keyspace'}')
rows = session.execute("""${sql}""")
for row in rows:
    print(row)

        `.trim();

      default:
        return ``;
    }
  }
}

export default new ConcreteDataScripts();
