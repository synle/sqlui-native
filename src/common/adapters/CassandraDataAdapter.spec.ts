import CassandraDataAdapter from 'src/common/adapters/CassandraDataAdapter';

describe(' 2 cassandra v4', () => {
  let adapter

  beforeAll(() => {
    adapter = new CassandraDataAdapter('cassandra://localhost:9042');
  });

  test('Get database', async () => {
    const databases = await adapter.getDatabases();
    expect(databases).toMatchInlineSnapshot(`
Array [
  Object {
    "name": "music_stores",
    "tables": Array [],
  },
  Object {
    "name": "system",
    "tables": Array [],
  },
  Object {
    "name": "system_auth",
    "tables": Array [],
  },
  Object {
    "name": "system_distributed",
    "tables": Array [],
  },
  Object {
    "name": "system_schema",
    "tables": Array [],
  },
  Object {
    "name": "system_traces",
    "tables": Array [],
  },
  Object {
    "name": "system_views",
    "tables": Array [],
  },
  Object {
    "name": "system_virtual_schema",
    "tables": Array [],
  },
]
`);
  });

  test('Get tables', async () => {
    const tables = await adapter.getTables('system');
    expect(tables).toMatchInlineSnapshot(`
Array [
  Object {
    "columns": Array [],
    "name": "available_ranges",
  },
  Object {
    "columns": Array [],
    "name": "available_ranges_v2",
  },
  Object {
    "columns": Array [],
    "name": "batches",
  },
  Object {
    "columns": Array [],
    "name": "built_views",
  },
  Object {
    "columns": Array [],
    "name": "compaction_history",
  },
  Object {
    "columns": Array [],
    "name": "IndexInfo",
  },
  Object {
    "columns": Array [],
    "name": "local",
  },
  Object {
    "columns": Array [],
    "name": "paxos",
  },
  Object {
    "columns": Array [],
    "name": "peer_events",
  },
  Object {
    "columns": Array [],
    "name": "peer_events_v2",
  },
  Object {
    "columns": Array [],
    "name": "peers",
  },
  Object {
    "columns": Array [],
    "name": "peers_v2",
  },
  Object {
    "columns": Array [],
    "name": "prepared_statements",
  },
  Object {
    "columns": Array [],
    "name": "repairs",
  },
  Object {
    "columns": Array [],
    "name": "size_estimates",
  },
  Object {
    "columns": Array [],
    "name": "sstable_activity",
  },
  Object {
    "columns": Array [],
    "name": "table_estimates",
  },
  Object {
    "columns": Array [],
    "name": "transferred_ranges",
  },
  Object {
    "columns": Array [],
    "name": "transferred_ranges_v2",
  },
  Object {
    "columns": Array [],
    "name": "view_builds_in_progress",
  },
]
`);
  });

  test('Get columns', async () => {
    const columns = await adapter.getColumns('columns', 'system_schema');
    expect(columns).toMatchInlineSnapshot(`
Array [
  Object {
    "kind": "regular",
    "name": "clustering_order",
    "type": "text",
  },
  Object {
    "kind": "clustering",
    "name": "column_name",
    "type": "text",
  },
  Object {
    "kind": "regular",
    "name": "column_name_bytes",
    "type": "blob",
  },
  Object {
    "kind": "partition_key",
    "name": "keyspace_name",
    "type": "text",
  },
  Object {
    "kind": "regular",
    "name": "kind",
    "type": "text",
  },
  Object {
    "kind": "regular",
    "name": "position",
    "type": "int",
  },
  Object {
    "kind": "clustering",
    "name": "table_name",
    "type": "text",
  },
  Object {
    "kind": "regular",
    "name": "type",
    "type": "text",
  },
]
`);
  });

  test('Execute Select', async () => {
    const resp = await adapter.execute(`SELECT * FROM tables LIMIT 10`, 'system_schema');
    //@ts-ignore
    expect(resp.raw.length > 0).toBeTruthy();
  });
});
