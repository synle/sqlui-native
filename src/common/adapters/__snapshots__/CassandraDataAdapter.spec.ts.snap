// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cassandra Get columns 1`] = `
Array [
  Object {
    "kind": "regular",
    "name": "clustering_order",
    "type": "text",
  },
  Object {
    "kind": "clustering",
    "name": "column_name",
    "type": "text",
  },
  Object {
    "kind": "regular",
    "name": "column_name_bytes",
    "type": "blob",
  },
  Object {
    "kind": "partition_key",
    "name": "keyspace_name",
    "type": "text",
  },
  Object {
    "kind": "regular",
    "name": "kind",
    "type": "text",
  },
  Object {
    "kind": "regular",
    "name": "position",
    "type": "int",
  },
  Object {
    "kind": "clustering",
    "name": "table_name",
    "type": "text",
  },
  Object {
    "kind": "regular",
    "name": "type",
    "type": "text",
  },
]
`;

exports[`cassandra Get database 1`] = `
Array [
  Object {
    "name": "music_stores",
    "tables": Array [],
  },
  Object {
    "name": "system",
    "tables": Array [],
  },
  Object {
    "name": "system_auth",
    "tables": Array [],
  },
  Object {
    "name": "system_distributed",
    "tables": Array [],
  },
  Object {
    "name": "system_schema",
    "tables": Array [],
  },
  Object {
    "name": "system_traces",
    "tables": Array [],
  },
  Object {
    "name": "system_views",
    "tables": Array [],
  },
  Object {
    "name": "system_virtual_schema",
    "tables": Array [],
  },
]
`;

exports[`cassandra Get tables 1`] = `
Array [
  Object {
    "columns": Array [],
    "name": "available_ranges",
  },
  Object {
    "columns": Array [],
    "name": "available_ranges_v2",
  },
  Object {
    "columns": Array [],
    "name": "batches",
  },
  Object {
    "columns": Array [],
    "name": "built_views",
  },
  Object {
    "columns": Array [],
    "name": "compaction_history",
  },
  Object {
    "columns": Array [],
    "name": "IndexInfo",
  },
  Object {
    "columns": Array [],
    "name": "local",
  },
  Object {
    "columns": Array [],
    "name": "paxos",
  },
  Object {
    "columns": Array [],
    "name": "peer_events",
  },
  Object {
    "columns": Array [],
    "name": "peer_events_v2",
  },
  Object {
    "columns": Array [],
    "name": "peers",
  },
  Object {
    "columns": Array [],
    "name": "peers_v2",
  },
  Object {
    "columns": Array [],
    "name": "prepared_statements",
  },
  Object {
    "columns": Array [],
    "name": "repairs",
  },
  Object {
    "columns": Array [],
    "name": "size_estimates",
  },
  Object {
    "columns": Array [],
    "name": "sstable_activity",
  },
  Object {
    "columns": Array [],
    "name": "table_estimates",
  },
  Object {
    "columns": Array [],
    "name": "transferred_ranges",
  },
  Object {
    "columns": Array [],
    "name": "transferred_ranges_v2",
  },
  Object {
    "columns": Array [],
    "name": "view_builds_in_progress",
  },
]
`;

exports[`cassandra v4 Get columns 1`] = `
Array [
  Object {
    "kind": "regular",
    "name": "clustering_order",
    "type": "text",
  },
  Object {
    "kind": "clustering",
    "name": "column_name",
    "type": "text",
  },
  Object {
    "kind": "regular",
    "name": "column_name_bytes",
    "type": "blob",
  },
  Object {
    "kind": "partition_key",
    "name": "keyspace_name",
    "type": "text",
  },
  Object {
    "kind": "regular",
    "name": "kind",
    "type": "text",
  },
  Object {
    "kind": "regular",
    "name": "position",
    "type": "int",
  },
  Object {
    "kind": "clustering",
    "name": "table_name",
    "type": "text",
  },
  Object {
    "kind": "regular",
    "name": "type",
    "type": "text",
  },
]
`;

exports[`cassandra v4 Get database 1`] = `
Array [
  Object {
    "name": "music_stores",
    "tables": Array [],
  },
  Object {
    "name": "system",
    "tables": Array [],
  },
  Object {
    "name": "system_auth",
    "tables": Array [],
  },
  Object {
    "name": "system_distributed",
    "tables": Array [],
  },
  Object {
    "name": "system_schema",
    "tables": Array [],
  },
  Object {
    "name": "system_traces",
    "tables": Array [],
  },
  Object {
    "name": "system_views",
    "tables": Array [],
  },
  Object {
    "name": "system_virtual_schema",
    "tables": Array [],
  },
]
`;

exports[`cassandra v4 Get tables 1`] = `
Array [
  Object {
    "columns": Array [],
    "name": "available_ranges",
  },
  Object {
    "columns": Array [],
    "name": "available_ranges_v2",
  },
  Object {
    "columns": Array [],
    "name": "batches",
  },
  Object {
    "columns": Array [],
    "name": "built_views",
  },
  Object {
    "columns": Array [],
    "name": "compaction_history",
  },
  Object {
    "columns": Array [],
    "name": "IndexInfo",
  },
  Object {
    "columns": Array [],
    "name": "local",
  },
  Object {
    "columns": Array [],
    "name": "paxos",
  },
  Object {
    "columns": Array [],
    "name": "peer_events",
  },
  Object {
    "columns": Array [],
    "name": "peer_events_v2",
  },
  Object {
    "columns": Array [],
    "name": "peers",
  },
  Object {
    "columns": Array [],
    "name": "peers_v2",
  },
  Object {
    "columns": Array [],
    "name": "prepared_statements",
  },
  Object {
    "columns": Array [],
    "name": "repairs",
  },
  Object {
    "columns": Array [],
    "name": "size_estimates",
  },
  Object {
    "columns": Array [],
    "name": "sstable_activity",
  },
  Object {
    "columns": Array [],
    "name": "table_estimates",
  },
  Object {
    "columns": Array [],
    "name": "transferred_ranges",
  },
  Object {
    "columns": Array [],
    "name": "transferred_ranges_v2",
  },
  Object {
    "columns": Array [],
    "name": "view_builds_in_progress",
  },
]
`;
